docs/changes/units/17796.bugfix.rst

```diff
@@ -0,0 +1,9 @@
+Ensure that ``CompositeUnit`` always have bases consisting of named units.
+A consequence of this is that decompositions into unnamed units such as
+``u.ks.decompose([u.Unit("100 s")])``, while now representable, will have
+different scale and bases (effectively, "1000 s" instead of "10 100 s",
+where the latter could not be represented as a standard unit).
+
+In practice, this should not affect anything, since the physical meaning of
+the unit has not changed.  If the scale is important, one should use named
+units (e.g., ``u.ks.decompose([u.Unit("mytime", "100 s")])``).
```

Comment:
Is the recommended way of creating a `NamedUnit` `Unit()` or `def_unit()`?

Reply:
I don't know that we have a true preferred method, but I see `u.Unit` as a quick way to define a throw-away unit, while `def_unit` is something more for in code. But the goal here was just to give the simplest possible change that would modify existing code and have it continue to run as intended.

Reply:
`Unit()` is fundamentally broken:
```python
>>> from astropy import units as u
>>> isinstance(u.Unit("m"), u.Unit)
False
>>> isinstance(u.Unit("km / h"), u.Unit)
False
```
Although I don't quite like how the `prefixes` parameter of `def_unit()` is annotated as https://github.com/astropy/astropy/blob/8bc991637215661e7e9d9431b6823d9e788730f6/astropy/units/core.py#L2586 instead of simply `prefixes: Iterable[UnitPrefix] = ()`, that is not a serious problem. So if we can use `def_unit()` instead of `Unit()` in user-facing documentation then we should.

Reply:
Agreed.